ntoskrnl/mm/section.c

```diff
@@ -4920,6 +4923,49 @@ MmPurgeSegment(
     return TRUE;
 }
 
+BOOLEAN
+NTAPI
+MmIsDataSectionResident(
+    _In_ PSECTION_OBJECT_POINTERS SectionObjectPointer,
+    _In_ LONGLONG Offset,
+    _In_ ULONG Length)
+{
+    PMM_SECTION_SEGMENT Segment;
+    LARGE_INTEGER RangeStart, RangeEnd;
+    BOOLEAN Ret = TRUE;
+
+    RangeStart.QuadPart = Offset;
+    if (!NT_SUCCESS(RtlLongLongAdd(RangeStart.QuadPart, Length, &RangeEnd.QuadPart)))
+        return FALSE;
+
+    Segment = MiGrabDataSection(SectionObjectPointer);
+    if (!Segment)
+        return FALSE;
+
+    /* Find start page offset */
+    RangeStart.QuadPart >>= PAGE_SHIFT;
+    RangeStart.QuadPart <<= PAGE_SHIFT;
```

Comment:
What's the purpose of this jingle?

Reply:
I've never used `PAGE_ROUND_DOWN` or `MM_ROUND_DOWN` before. The purpose of this code is "round down" the byte offset to the offset of the page to start. For example, offset _0-4095 -> 0_ or offset _4096-8191 -> 4096_ and offset _8192-12287 -> 8192_.

Reply:
Replaced with the `PAGE_ROUND_DOWN` macro.